### PR TITLE
feat(tier4_autoware_utils): add trajectory func

### PR DESCRIPTION
**Pull request by yn-mrse**
## Description

下記の関数を現時点autoware.universe 最新の https://github.com/tier4/autoware.universe/blob/514fea4bf80bf460fa880d96377eb1230823fb7b/common/motion_utils/include/motion_utils/trajectory/trajectory.hpp からコピーします。
- validateNonSharpAngle
- isDrivingForward
- isDrivingForwardWithTwist
- insertTargetPoint

また、現時点autoware.universe最新の https://github.com/tier4/autoware.universe/blob/7ef1bee35c60cfc4637840bfacc0162d14e31143/common/motion_utils/include/motion_utils/trajectory/path_with_lane_id.hpp をコピーし、tier4_autoware_utilsとして置き換えます。
## Related links

- https://github.com/tier4/autoware.universe/blob/514fea4bf80bf460fa880d96377eb1230823fb7b/common/motion_utils/include/motion_utils/trajectory/trajectory.hpp#L45-L117
- https://github.com/tier4/autoware.universe/blob/514fea4bf80bf460fa880d96377eb1230823fb7b/common/motion_utils/include/motion_utils/trajectory/trajectory.hpp#L968-L1186
- https://github.com/tier4/autoware.universe/blob/7ef1bee35c60cfc4637840bfacc0162d14e31143/common/motion_utils/include/motion_utils/trajectory/path_with_lane_id.hpp

## Tests performed

- trajectory.hpp内の関数は、上記リンク箇所の実装と完全に同一内容であることを確認済みです。
- path_with_lane_id.hpp内実装は、上記リンク箇所の実装と同一であり、"motion_utils"だけが"tier4_autoware_utils"へと変更になっていることを確認済みです。
- build testはhttps://github.com/tier4/autoware.universe/pull/596 にてまとめて確認済みです。

## Pre-review checklist for the PR author

The PR author **must** check the checkboxes below when creating the PR.

- [x] I've confirmed the [contribution guidelines].
- [x] The PR follows the [pull request guidelines].

## In-review checklist for the PR reviewers

The PR reviewers **must** check the checkboxes below before approval.

- [ ] The PR follows the [pull request guidelines].
- [ ] The PR has been properly tested.
- [ ] The PR has been reviewed by the code owners.

## Post-review checklist for the PR author

The PR author **must** check the checkboxes below before merging.

- [ ] There are no open discussions or they are tracked via tickets.
- [ ] The PR is ready for merge.

After all checkboxes are checked, anyone who has write access can merge the PR.

[contribution guidelines]: https://autowarefoundation.github.io/autoware-documentation/main/contributing/
[pull request guidelines]: https://autowarefoundation.github.io/autoware-documentation/main/contributing/pull-request-guidelines/

### CODE REVIEW

File: common/tier4_autoware_utils/include/tier4_autoware_utils/trajectory/path_with_lane_id.hpp [hunk 1]
```diff
@@ -0,0 +1,113 @@
...
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef TIER4_AUTOWARE_UTILS__TRAJECTORY__PATH_WITH_LANE_ID_HPP_
+#define TIER4_AUTOWARE_UTILS__TRAJECTORY__PATH_WITH_LANE_ID_HPP_
+
+#include "motion_utils/trajectory/trajectory.hpp"
```
Comment (yn-mrse) [inline common/tier4_autoware_utils/include/tier4_autoware_utils/trajectory/path_with_lane_id.hpp:18]:
```suggestion
#include "tier4_autoware_utils/trajectory/trajectory.hpp"
```

File: common/tier4_autoware_utils/include/tier4_autoware_utils/trajectory/path_with_lane_id.hpp [hunk 2]
```diff
@@ -0,0 +1,113 @@
...
+#include "tier4_autoware_utils/trajectory/trajectory.hpp"
+#include "tier4_autoware_utils/geometry/path_with_lane_id_geometry.hpp"
+
+#include <boost/optional.hpp>
+
+#include <algorithm>
+#include <utility>
+#include <vector>
+
+namespace motion_utils
```
Comment (yn-mrse) [inline common/tier4_autoware_utils/include/tier4_autoware_utils/trajectory/path_with_lane_id.hpp:27]:
```suggestion
namespace tier4_autoware_utils
```

File: common/tier4_autoware_utils/include/tier4_autoware_utils/trajectory/path_with_lane_id.hpp [hunk 3]
```diff
@@ -0,0 +1,113 @@
...
+
+// @brief Calculates the path to be followed by the rear wheel center in order to make the vehicle
+// center follow the input path
+// @param [in] path with position to be followed by the center of the vehicle
+// @param [out] PathWithLaneId to be followed by the rear wheel center follow to make the vehicle
+// center follow the input path NOTE: rear_to_cog is supposed to be positive
+autoware_auto_planning_msgs::msg::PathWithLaneId convertToRearWheelCenter(
+  const autoware_auto_planning_msgs::msg::PathWithLaneId & path, const double rear_to_cog,
+  const bool enable_last_point_compensation = true);
+}  // namespace motion_utils
```
Comment (yn-mrse) [inline common/tier4_autoware_utils/include/tier4_autoware_utils/trajectory/path_with_lane_id.hpp:111]:
```suggestion
}  // namespace tier4_autoware_utils
```

**Comment by sfukuta** [comment]
Related linksにある
https://github.com/tier4/autoware.universe/blob/514fea4bf80bf460fa880d96377eb1230823fb7b/common/motion_utils/include/motion_utils/trajectory/trajectory.hpp#L1220-L1252
の"insertStopPoint"が修正内容に存在していないです。
Descriptionともあっていないので、どちらか修正をお願いします。

**Review by sfukuta — APPROVED** [review]

**Pull request merged**